### PR TITLE
use extractall(filter=) Python 3.12+ feature

### LINKED ISSUE #87: `tarfile` filters `DeprecationWarning`
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

conda-build's CI is plagued with warnings like this:

```
DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.
```

conda-package-streaming should start adding the filter option explicitly for Python 3.12+.

### Conda Info

```shell
N/A
```

### Conda Config

```shell
N/A
```

### Conda list

```shell
N/A
```

### Additional Context

See https://github.com/conda/conda-build/pull/5281

### PR DESCRIPTION
**Pull request by dholth**
### Description

Fix #87 

Code coverage testing might be tricky on this one?

Did we pick the correct filter? Can we backport it ourselves to older Python, since the Python filter is more sophisticated than our internal version?

(The tests are failing because data_filter _has_ been backported to Python < 3.12)

### CODE REVIEW

**Review by jezdez — DISMISSED** [review]

**Comment by jezdez** [comment]
I'm not sure I understand why the tests fail..

File: conda_package_streaming/extract.py [hunk 1]
```diff
@@ -44,7 +45,14 @@ def checked_members():
                 yield member
 
         try:
-            tar_file.extractall(path=dest_dir, members=checked_members())
+            if HAS_TAR_FILTER:
+                tar_file.extractall(
+                    path=dest_dir,
+                    members=checked_members(),
+                    filter=tarfile.fully_trusted_filter,
```
Comment (jezdez) [inline conda_package_streaming/extract.py:52]: From reading https://docs.python.org/3/library/tarfile.html#tarfile-extraction-filter closely:
> The filter argument to [TarFile.extract()](https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extract) or [extractall()](https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractall) can be:
>
> - the string 'fully_trusted': Honor all metadata as specified in the archive. Should be used if the user trusts the archive completely, or implements their own complex verification.
>
> - the string 'tar': Honor most tar-specific features (i.e. features of UNIX-like filesystems), but block features that are very likely to be surprising or malicious. See [tar_filter()](https://docs.python.org/3/library/tarfile.html#tarfile.tar_filter) for details.
>
> - the string 'data': Ignore or block most features specific to UNIX-like filesystems. Intended for extracting cross-platform data archives. See [data_filter()](https://docs.python.org/3/library/tarfile.html#tarfile.data_filter) for details.
>
> - None (default): Use [TarFile.extraction_filter](https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extraction_filter).

```suggestion
                    filter="fully_trusted",
```

I'm aware that `_get_filter_function` will check for the callable internally, but that's an implementation detail and we should use the standard API.

**Review by jezdez — COMMENTED** [review]
I'm wondering if we should adopt tar_filter, but I may be not fully aware what other limits it implements, my POSIX foo is pretty weak.

https://github.com/python/cpython/blob/54a05a46002ee1c9211f299df38f444f16866ef5/Lib/tarfile.py#L769-L772 seems pretty close to our `checked_members()` function though.

File: conda_package_streaming/extract.py (hunk 1 above)
Comment (dholth) [reply conda_package_streaming/extract.py:52]: stringly typed API, yuck.

File: conda_package_streaming/extract.py [hunk 2]
```diff
@@ -44,7 +45,14 @@ def checked_members():
...
+                )
+            else:
+                tar_file.extractall(path=dest_dir, members=checked_members())
```
[the remark this answers is not in the document]
Comment (kenodegard) [reply conda_package_streaming/extract.py:55]: or alternatively:

```python
kwargs = {"filter": "fully_trusted"} if HAS_TAR_FILTER else {}

try:
    tar_file.extractall(
        ...,
        **kwargs,
    )
...
```

[the remark this answers is not in the document]
Comment (dholth) [reply conda_package_streaming/extract.py:55]: Combining fully_trusted with our existing filter is the least risky option.

File: conda_package_streaming/package_streaming.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def __str__(self):
 
 
 class TarfileNoSameOwner(tarfile.TarFile):
-    def __init__(self, *args, umask=UMASK, **kwargs):
```
Comment (dholth) [inline conda_package_streaming/package_streaming.py:39]: Is this change causing our test to fail?

**Comment by dholth** [comment]
I've concluded that the umask handling is causing most of the test failures here (where we attempt to grab umask on the fly instead of as a default function argument). Will separate that change and the tar-filter bit into two PR's

**Comment by dholth** [comment]
Now that the tests seem to have been worked out, we could use a better filter than fully_trusted...

**Review by jjhelmus — APPROVED** [review]

**Review by jezdez — APPROVED** [review]

**Comment by dholth** [comment]
Thanks

**Pull request merged**